proposals/NNNN-android-transport-overhaul.md

```diff
@@ -0,0 +1,491 @@
+
+The current USB and TCP transport classes are very fragile and has a tendency to deadlock and prevent reconnection. A new USB and TCP transport will be created that uses an also new `MultiplexBaseTransport` abstract class [(_Code Snippet_)](#multiplexbasetransport-code). The new `MultiplexUsbTransport` [(_Code Snippet_)](#multiplexusbtransport-code) and `MultiplexTCPTransport` [(_Code Snippet_TBD_)](#TBD)will be greatly simplified.
+
+
+
+#### MultiplexBluetoothTransport changes
+ The `MultiplexBluetoothTransport` will also now extend from the `MultiplexBaseTransport`
+ 
+
+#### Handle multiple transports per App per Session
+
+Each bound app will maintain one binding reference and the router service will handle when transports are added to the `RegisteredApp` and which session specifically.
+
+```java
+//RegisteredApp class
+
+//Primary will always be first. Key is the session id
+SparseArray<ArrayList<TransportType>> registeredTransports;
+
+protected void registerTransport(int sessionId,TransportType transportType){
+	synchronized (TRANSPORT_LOCK){
+		ArrayList<TransportType> transportTypes = this.registeredTransports.get(sessionId);
+		if(transportTypes!= null){
+			if(queues.get(transportType) == null){
+				queues.put(transportType, new PacketWriteTaskBlockingQueue());
+			}
+			transportTypes.add(transportType);
+			this.registeredTransports.put(sessionId,transportTypes);
+		}
+	}
+}
+
+protected boolean isRegisteredOnTransport(int sessionId, TransportType transportType){
+	synchronized (TRANSPORT_LOCK){
+		if(this.registeredTransports.indexOfKey(sessionId) >= 0){
+			return this.registeredTransports.get(sessionId).indexOf(transportType) >= 0;
+		}else{
+			return false;
+		}
+	}
+}
+
+protected List<TransportType> getTransportsForSession(int sessionId){
+	synchronized (TRANSPORT_LOCK){
+		if(this.registeredTransports.indexOfKey(sessionId) >= 0){
+			return this.registeredTransports.get(sessionId);
+		}else{
+			return null;
+		}
+	}
+}
+
+protected boolean unregisterTransport(int sessionId, TransportType transportType){
+	synchronized (TRANSPORT_LOCK){
+		if(this.registeredTransports.indexOfKey(sessionId) >= 0){
+			return this.registeredTransports.get(sessionId).remove(transportType);
+		}else{
+			return false;
+		}
+	}
+}
+
+protected void unregisterAllTransports(int sessionId){
+	synchronized (TRANSPORT_LOCK){
+		if(this.registeredTransports.indexOfKey(sessionId) >= 0){
+			this.registeredTransports.get(sessionId).clear();
+		}else if(sessionId == -1){
+			int size = this.registeredTransports.size();
+			for(int i = 0; i <size; i++){
+				this.registeredTransports.get(i).clear();
+			}
+		}
+	}
+}
+
+```
+
+#### Packet Queues
+Packet queues will be a per transport per app item. Each bound app will have a queue per their registered transport for all sessions to operate.
+
+```java
+HashMap<TransportType, PacketWriteTaskBlockingQueue> queues;
+
+public boolean handleIncommingClientMessage(final Bundle receivedBundle){
```

Comment:
Fix typo "handleIncomingClientMessage"

Reply:
This is already implemented into the project and would cause a breaking change. Wouldn't be worth it.